web/source/keyboards/kmwkeyboards.ts

```diff
@@ -1143,31 +1143,65 @@ namespace com.keyman.keyboards {
      *  @param  {string}   cmd        command string
      *  @param  {boolean?} byLanguage if true, context=languages, else context=keyboards
      **/
-    keymanCloudRequest(cmd: string, byLanguage?: boolean) {
-      var URL='https://api.keyman.com/cloud/4.0/', tFlag,
-        Lscript = this.keymanweb.util._CreateElement('script');
-
-      URL = URL + ((arguments.length > 1) && byLanguage ? 'languages' : 'keyboards')
-        +'?jsonp=keyman.register&languageidtype=bcp47&version='+this.keymanweb['version'];
-
+    keymanCloudRequest(cmd: string, byLanguage?: boolean) {    
       var kbdManager = this;
+      var keymanweb = this.keymanweb;
+
+      var TIMEOUT_KEY = "cloud request timed out";
+
+      // Some basic support toward #5044, but definitely not a full solution toward it.
+      // We'd want to define an 'associative array' for registration promises, using
+      // the timer ID as the 'key' that would help us retrieve the matching Promise.
+      //
+      // Given that, if the Promise were wrapped and `resolve` and `reject` captured,
+      // the `register` function could then forward the succesfully-retrieved stubs to
+      // the `resolve` method upon success.
+      let promise = new Promise<void>(function(resolve, reject) {
+        var URL='https://api.keyman.com/cloud/4.0/';
+        var tFlag: string;
+        let Lscript: HTMLScriptElement = keymanweb.util._CreateElement('script');
+
+        URL = URL + ((arguments.length > 1) && byLanguage ? 'languages' : 'keyboards')
+          +'?jsonp=keyman.register&languageidtype=bcp47&version='+keymanweb['version'];
+  
+        // Set callback timer
+        let timeoutID = window.setTimeout(function() {
+          reject(TIMEOUT_KEY);
+        } ,10000);
+
+        tFlag='&timerid='+ timeoutID;
+  
+        Lscript.charset="UTF-8";
+        Lscript.src = URL+cmd+tFlag;
```

Comment:
We should set `src` last, and apparently should set `onload` after appending too. I haven't yet found an authoritative reference for this advice though!

https://stackoverflow.com/a/16231055/1836776

And, to fit with scoping changes I made to `URL`:
```suggestion
        Lscript.src = URL+urlParams+cmd+tFlag;
```

Reply:
Oh right, I forgot about that.  Very true.